Sources/NIOWebSocketClient/main.swift

```diff
@@ -0,0 +1,233 @@
+    typealias InboundIn = WebSocketFrame
+    typealias OutboundOut = WebSocketFrame
+    
+    let testFrameData: String = "Hello World"
+    
+    // This is being hit, channel active won't be called as it is already added.
+    public func handlerAdded(context: ChannelHandlerContext) {
+        print("WebSocket handler added.")
+        self.pingTestFrameData(context: context)
+    }
+
+    public func handlerRemoved(context: ChannelHandlerContext) {
+        print("WebSocket handler removed.")
+    }
+
+    public func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        let frame = self.unwrapInboundIn(data)
+        
+        switch frame.opcode {
+        case .pong:
+            self.pong(context: context, frame: frame)
+        case .text:
+            var data = frame.unmaskedData
+            let text = data.readString(length: data.readableBytes) ?? ""
+            print("Websocket: Received \(text)")
+        case .connectionClose:
+            self.receivedClose(context: context, frame: frame)
+        case .binary, .continuation, .ping:
+            // We ignore these frames.
+            break
+        default:
+            // Unknown frames are errors.
+            self.closeOnError(context: context)
+        }
+    }
+    
+    public func channelReadComplete(context: ChannelHandlerContext) {
+        context.flush()
+    }
+
+    private func receivedClose(context: ChannelHandlerContext, frame: WebSocketFrame) {
+        // Handle a received close frame. We're just going to close.
+        print("Received Close instruction from server")
+        context.close(promise: nil)
+    }
+    
+    private func pingTestFrameData(context: ChannelHandlerContext) {
+        let frameData = ByteBuffer(string: self.testFrameData)
+        let frame = WebSocketFrame(fin: true, opcode: .ping, data: frameData)
+        context.write(self.wrapOutboundOut(frame), promise: nil)
+    }
+    
+    private func pong(context: ChannelHandlerContext, frame: WebSocketFrame) {
+        if let frameData = frame.data.getString(at: 0, length: self.testFrameData.count) {
+            print("Websocket: Received: \(frameData)")
+        }
+    }
+    
+    private func closeOnError(context: ChannelHandlerContext) {
+        // We have hit an error, we want to close. We do that by sending a close frame and then
+        // shutting down the write side of the connection. The server will respond with a close of its own.
+        var data = context.channel.allocator.buffer(capacity: 2)
+        data.write(webSocketErrorCode: .protocolError)
+        let frame = WebSocketFrame(fin: true, opcode: .connectionClose, data: data)
+        context.write(self.wrapOutboundOut(frame)).whenComplete { (_: Result<Void, Error>) in
+            context.close(mode: .output, promise: nil)
+        }
+    }
+}
+
+extension ByteBuffer {
+    fileprivate init(string: String) {
+        self = ByteBufferAllocator().buffer(capacity: string.utf8.count)
```

Comment:
As these are intended to be examples of proper NIO usage, can we use the `ByteBufferAllocator` from the `Channel`?